### PR TITLE
Fix multiselect with a lot of options not saved to DB

### LINKED ISSUE #33486: Multiselect attribute, not all options are saved in database.
### Preconditions (*)

1. Magento 2.4.2, 
2. Magento 2.4.1, 
3. Magento 2.4.2-p1, 
4. Magento 2.3.x

### Steps to reproduce (*)

1. Create a new `multiselect` attribute
2. Add at least **150** options to an attribute
3. Add an attribute to main attribute set
4. create a product and assign ALL options to the product 

### Expected result (*)

1. All options are selected (as they are not saved) in `catalog_product_entity_text` table as in Magento 1 (or propose a different solution)

### Actual result (*)

1. Attribute options are saved in `catalog_produdct_entity_varchar` table 
2. not all options are saved when you reopen the product in Admin

This is basically a copy of this issue:
https://github.com/magento/magento2/issues/24199

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [x] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by jakwinkler**
### Description (*)
Code fix to https://github.com/magento/magento2/issues/33486

### Fixed Issues (if relevant)

1. Fixes magento/magento2#33486

### Manual testing scenarios (*)

1. Create a new multiselect attribute
2. Add at least 150 options to an attribute
3. Add an attribute to the main attribute set
4. create a product and assign ALL options to the product
5. After save operation, all values should be assigned to the product

### Questions or comments

1. Who tested this feature and why it was tested on very small data set?

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by qsolutions-pl** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
That will work for newly created installation, but will not for older. Could you create some patch that will migrate all data from varchar to text table?

Also, could you create some integration test for covering this case?

**Comment by ihor-sviziev** [comment]
@sivaschenko @sidolov @gabrieldagama, I think the prio should be P2 for this issue.

**Comment by qsolutions-pl** [comment]
@ihor-sviziev added integration tests :-)

**Comment by qsolutions-pl** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: dev/tests/integration/testsuite/Magento/Catalog/_files/multiselect_attribute.php [hunk 1]
```diff
@@ -34,21 +43,11 @@
...
-                    'option_3' => ['Option 3'],
-                    'option_4' => ['Option 4 "!@#$%^&*']
-                ],
-                'order' => [
-                    'option_1' => 1,
-                    'option_2' => 2,
-                    'option_3' => 3,
-                    'option_4' => 4,
-                ],
+                'value' => $valueOptionArray,
```
Comment (ihor-sviziev) [inline dev/tests/integration/testsuite/Magento/Catalog/_files/multiselect_attribute.php:101]: Please revert changing of the values. Instead, please just copy-paste this file and the one with the _rollback.php suffix and adjust it how you need it.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by qsolutions-pl** [comment]
Since this is a critical issue, it affects the entire framework functionality for products with multi select attributes and it got NONE attention from M2 CORE TEAM, I'm closing this issue and related Pull Request.

**Comment by Den4ik** [comment]
Hi @qsolutions-pl, you were received request for change but didn't respond. Sorry that you didn't receive any response, but "need updates" label mean that PR checked by someone of maintainers and require updates or response from PR author.

PS. I don't see migration of exists options from varchar to text table

**Comment by qsolutions-pl** [comment]
@Den4ik I tried to create a new test just for the changes I've done in the PR.  However, when I created a new test that adds a new multiselect attribute, this new attribute was not saved in the database and the new test was failing. Code of the test, data fixtures were basically a copy-paste of the `ProductGettersTest.php` with few simple modifications. 
I and @bgorski tried to solve this on 28th August and were not able to do this. 
You can see the new test here: 
https://github.com/qsolutions-pl/magento2/commit/de2e7f7e021f08dccce5984a6983efc0b5e9655f 
I've even tried to comment out native tests in `ProductGettersTest` file to see if my new test will run, but no luck. 
I believe it has to do something with `repository cache` but that is just a guess.
I haven't pushed the migration script to the repository yet, but it is here:
https://github.com/wmwnuk/magento2/commit/767987c30f4aa0043a7eac1ddc08b9149a95b6e7 as I've assigned one of my team members to do it while I tried to fix tests at the same time.

**Comment by Den4ik** [comment]
@qsolutions-pl Could you please prepare new one PR that contain tests and migration script and I'll try to help if tests will be failed

**Comment by qsolutions-pl** [comment]
Reopening PR

**Comment by qsolutions-pl** [comment]
@magento run all tests

File: dev/tests/integration/testsuite/Magento/Catalog/_files/multiselect_attribute.php (hunk 1 above)
Comment (jakwinkler) [reply dev/tests/integration/testsuite/Magento/Catalog/_files/multiselect_attribute.php:101]: I've modified fixtures to work with 2 multiselect attributes, you can mark this as resolved

**Comment by qsolutions-pl** [comment]
This test will fail:
`testsuite/Magento/Catalog/Model/ResourceModel/Attribute/Entity/AttributeTest.php`

**Comment by qsolutions-pl** [comment]
@magento run all tests

**Comment by bgorski** [comment]
@magento run all tests

**Comment by qsolutions-pl** [comment]
@magento run all tests

**Comment by bgorski** [comment]
@magento run all tests

**Comment by qsolutions-pl** [comment]
@magento run all tests

File: dev/tests/integration/testsuite/Magento/Catalog/Model/ResourceModel/Product/Indexer/Eav/SourceTest.php [hunk 2]
```diff
@@ -14,11 +14,6 @@
 use Magento\Store\Model\StoreManagerInterface;
 use Magento\Store\Api\Data\StoreInterface;
 
-/**
- * Class SourceTest
- * @magentoAppIsolation enabled
```
Comment (ihor-sviziev) [inline dev/tests/integration/testsuite/Magento/Catalog/Model/ResourceModel/Product/Indexer/Eav/SourceTest.php:19]: I guess this should be retuned back

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: dev/tests/integration/testsuite/Magento/Catalog/Model/ResourceModel/Product/Indexer/Eav/SourceTest.php (hunk 2 above)
Comment (jakwinkler) [reply dev/tests/integration/testsuite/Magento/Catalog/Model/ResourceModel/Product/Indexer/Eav/SourceTest.php:19]: Static tests does not pass with it :-)

Comment (jakwinkler) [reply dev/tests/integration/testsuite/Magento/Catalog/Model/ResourceModel/Product/Indexer/Eav/SourceTest.php:19]: When I leave `Class SourceTest` it does not pass static tests
image

So I left it this way:
image

**Comment by qsolutions-pl** [comment]
@magento run all tests

**Comment by qsolutions-pl** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

**Review by ihor-sviziev — APPROVED** [review]
✔ Approved.

Failing tests look not related to changes from this PR.

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9225](https://jira.corp.magento.com/browse/ENGCOM-9225) has been created to process this Pull Request

**Comment by qsolutions-pl** [comment]
@magento run Functional Tests B2B

**Comment by qsolutions-pl** [comment]
@magento run Functional Tests B2B

**Review by Den4ik — APPROVED** [review]
✅ @qsolutions-pl Looks like very good job

**Comment by magento-engcom-team** [comment]
Hi @Den4ik, thank you for the review.
[ENGCOM-9225](https://jira.corp.magento.com/browse/ENGCOM-9225) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

1. Magento latest instance installed with sample data


**Manual testing scenario:**

1. Create a new attribute type multiselect, and add ~150 options to the attribute.

2. Create a new attribute set and use the above created attribute into this attribute set.
 
3. Create a new product by selecting the above created attribute set (Select all the attribute options while creating a new product.)

4. View the same created product in the same product created page.


**Before:** :heavy_multiplication_x: All the attribute options were not getting selected after creating a product

image


**After:** :heavy_check_mark: All the attribute options are getting selected after creating a product

image



Since it is a specific case and not impacting any other feature overall, there is no additional regression is required!

**Review by bgorski — APPROVED** [review]

**Comment by sivaschenko** [comment]
@qsolutions-pl it is currently in the "Merge in progress" status and should be merged following the queue among other high priority pull requests

**Pull request merged**